### PR TITLE
Migrate core devicelab framework to null safety.

### PR DESCRIPTION
**Pull request by darrenaustin**
Part of #84014.

This migrates the core framework parts of the devicelab code. It doesn't turn on sound null safety for running the tasks yet. It just allows other parts of the devicelab code to run with null safety when they are migrated.

### CODE REVIEW

File: dev/devicelab/lib/framework/browser.dart [hunk 1]
```diff
@@ -232,15 +230,12 @@ String _findSystemChromeExecutable() {
     return '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome';
   } else if (io.Platform.isWindows) {
     const String kWindowsExecutable = r'Google\Chrome\Application\chrome.exe';
-    final List<String> kWindowsPrefixes = <String>[
+    final List<String> kWindowsPrefixes = <String?>[
       io.Platform.environment['LOCALAPPDATA'],
       io.Platform.environment['PROGRAMFILES'],
       io.Platform.environment['PROGRAMFILES(X86)'],
-    ];
+    ].whereType<String>().toList();
```
Comment (justinmc) [inline dev/devicelab/lib/framework/browser.dart:237]: Ah clever way to remove nulls, I hadn't thought of that before.

File: dev/devicelab/lib/framework/browser.dart [hunk 2]
```diff
@@ -488,19 +483,19 @@ class BlinkTraceEvent {
   final String ph;
 
   /// Process ID of the process that emitted the event.
-  final int pid;
+  final int? pid;
```
Comment (justinmc) [inline dev/devicelab/lib/framework/browser.dart:486]: These parameters are all required but still nullable?

File: dev/devicelab/lib/framework/devices.dart [hunk 3]
```diff
@@ -568,18 +564,18 @@ class AndroidDevice extends Device {
...
     return <String, dynamic>{
-      'total_kb': int.parse(match.group(1)),
+      'total_kb': int.parse(match!.group(1)!),
     };
   }
 
   @override
   bool get canStreamLogs => true;
 
   bool _abortedLogging/*!*/ = false;
```
Comment (justinmc) [inline dev/devicelab/lib/framework/devices.dart:577]: What's up with these commented out `!`s and `?`s?  Notes for migration to null safety or something?  Should this one be removed?

File: dev/devicelab/lib/framework/ios.dart [hunk 4]
```diff
@@ -42,17 +40,10 @@ Future<bool> containsBitcode(String pathToBinary) async {
...
         .take(4)
-        .firstWhere(
-          (String line) => line.contains(' size 0x0000000000000001'),
-          orElse: () => null,
-      );
-      if (emptyBitcodeMarker != null) {
-        emptyBitcodeMarkerFound = true;
-        return;
-      }
+        .any((String line) => line.contains(' size 0x0000000000000001'));
```
Comment (justinmc) [inline dev/devicelab/lib/framework/ios.dart:46]: Nice, this is more slick 👍

**Review by justinmc — APPROVED** [review]
LGTM, but take a look at my comments in case I caught anything small.

File: dev/devicelab/lib/framework/devices.dart (hunk 3 above)
Comment (darrenaustin) [reply dev/devicelab/lib/framework/devices.dart:577]: Yeah, it looks like some of this was at one time ported in some fashion to null safety but then commented out. I got most of them, but missed this one. I will fix it. Thanks.

File: dev/devicelab/lib/framework/browser.dart (hunk 1 above)
Comment (darrenaustin) [reply dev/devicelab/lib/framework/browser.dart:237]: I stole it from @jmagman's review of another null safety migration.

File: dev/devicelab/lib/framework/browser.dart (hunk 2 above)
Comment (darrenaustin) [reply dev/devicelab/lib/framework/browser.dart]: It looks like some changes I meant to go in got lost in the shuffle here. I updated it with my intended changes. They should be non-nullable now.

**Comment by Hixie** [comment]
test-exempt: refactor with no semantic change

**Comment by whesse** [comment]
When running benchmarks on triple-headed builds (tip-of-tree of framework, engine, and Dart SDK), the gitBranch parameter is null.
Upload to cocoon is a separate step in the drone recipe, that only runs if an builder property is set, and so only runs when this is non-null.  But there are runs of the drone recipe that only upload the results file to CAS, with a null gitBranch (CommitBranch in the json).

So: the assert non-null in cocoon.dart:118 is incorrect, but was never run in production
The gitBranch parameter must be nullable in  writeTaskResultToFile

The check of gitBranch for non-null added in line 45 should not be added.

An example of a build uploading json with a null value for CommitBranch is
https://luci-milo.appspot.com/ui/p/flutter/builders/try/Linux_staging%20hot_mode_dev_cycle_linux__benchmark/232
and the uploaded json can be seen through the CAS UI on the upload to cas step.

**Pull request merged**